### PR TITLE
 Sct Events: Prometheus/Grafana Dumper

### PR DESCRIPTION
**Pull request by fruch**
New dashboard and grafana annotations based on SCT Events information

Annotations are backed in annotations.json file in S3, so we later can put them into grafana when needed:
    
```bash
./sct.py investigate show-logs b244db2a-346e-424c-8373-c50c3730fe2e
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|                                                Log links for testrun with test id b244db2a-346e-424c-8373-c50c3730fe2e                                                |
+-------------+---------------------------------------------------------------------------------------------------------------------------------------------------------+
| Log type    | Link                                                                                                                                                    |
+-------------+---------------------------------------------------------------------------------------------------------------------------------------------------------+
| annotations | https://cloudius-jenkins-test.s3.amazonaws.com/b244db2a-346e-424c-8373-c50c3730fe2e/annotations.json                                                    |
| grafana     | https://cloudius-jenkins-test.s3.amazonaws.com/b244db2a-346e-424c-8373-c50c3730fe2e/grafana-screenshot-overview-metrics-20190417_094331-0.png           |
| grafana     | https://cloudius-jenkins-test.s3.amazonaws.com/b244db2a-346e-424c-8373-c50c3730fe2e/grafana-screenshot-per-server-metrics-nemesis-20190417_094212-0.png |
| job         | https://cloudius-jenkins-test.s3.amazonaws.com/b244db2a-346e-424c-8373-c50c3730fe2e/job-2019-04-17T07.24-1173882.zip                                    |
| db-cluster  | https://cloudius-jenkins-test.s3.amazonaws.com/b244db2a-346e-424c-8373-c50c3730fe2e/longevity-fruch-fruch-db-cluster-b244db2a.zip                       |
| monitor-set | https://cloudius-jenkins-test.s3.amazonaws.com/b244db2a-346e-424c-8373-c50c3730fe2e/longevity-fruch-fruch-monitor-set-b244db2a.zip                      |
| prometheus  | https://cloudius-jenkins-test.s3.amazonaws.com/b244db2a-346e-424c-8373-c50c3730fe2e/prometheus_snapshot-b244db2a.tar.gz                                 |
+-------------+---------------------------------------------------------------------------------------------------------------------------------------------------------+
```

### CODE REVIEW

**Comment by fruch** [comment]
thanks @juliayakovlev for you time reviewing the dashboard, here are the things I've taken out of it:
* [ ] change the table to regular background (i.e. not transparent)
* [ ] Add Split to 3 filters, NORMAL / ERORR / CRITIAL 
* [ ] Check if we can add more of the tags to the filter in a dropdown or something like that
* [ ] in DbLogEvent move the regex closer to the type of the error
* [ ] Add the types of event (i.e. name of Nemesis, or DbLogEvent) as tags

File: sdcm/cluster.py [hunk 1]
```diff
@@ -3309,6 +3341,8 @@ def get_grafana_screenshot_and_snapshot(self, test_start_time=None):
         ]
 
         screenshot_url_tmpl = "http://{node_ip}:{grafana_port}/dashboard/db/{scylla_pkg}-{scr_name}-{version}?from={st}&to=now"
+
+        annotations_url = "http://{node_ip}:{grafana_port}/api/annotations"
```
Comment (bentsi) [inline sdcm/cluster.py]: remove

File: sdcm/cluster.py [hunk 2]
```diff
@@ -3331,7 +3365,14 @@ def get_grafana_screenshot_and_snapshot(self, test_start_time=None):
                     screenshots.append(self._get_screenshot_link(grafana_url, snapshot_path))
                     snapshots.append(self._get_shared_snapshot_link(grafana_url))
 
-                return {'screenshots': screenshots, 'snapshots': snapshots}
+                annotations = self.get_grafana_annotations(node)
+                if annotations:
```
Comment (bentsi) [inline sdcm/cluster.py]: separate to specific fucntion

File: sdcm/sct_events.py [hunk 3]
```diff
@@ -122,20 +130,69 @@ class Severity(enum.Enum):
     ERROR = 3
     CRITICAL = 4
 
+# TODO: move to utils
```
Comment (bentsi) [inline sdcm/sct_events.py]: move to utils

File: sdcm/sct_events.py [hunk 4]
```diff
@@ -122,20 +130,69 @@ class Severity(enum.Enum):
...
+
 
 class SctEvent(object):
     def __init__(self):
         self.timestamp = self.set_timestamp()
         self.severity = Severity.NORMAL
 
     def set_timestamp(self):
-        return datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S.%f")[:-3]
+        return time.time()
```
Comment (bentsi) [inline sdcm/sct_events.py]: consider removing

File: sdcm/sct_events.py [hunk 5]
```diff
@@ -280,8 +337,8 @@ def add_info_and_publish(self, node, line, line_number):
         self.publish()
 
     def __str__(self):
-        return "{}: type={} node={} line_number={} regex={}\n{}".format(
-            super(DatabaseLogEvent, self).__str__(), self.type, self.node, self.line_number, self.regex, self.line)
+        return "{}: type={} regex={} line_number={} node={}  \n{}".format(
```
Comment (bentsi) [inline sdcm/sct_events.py]: "{0}: type={1.type} regex={1.regex} line_number={} node={}  \n{}".format(str(self), self)

File: sdcm/sct_events.py [hunk 6]
```diff
@@ -331,30 +388,117 @@ def run(self):
...
         for event_type, message_data in EVENTS_PROCESS.subscribe_events():
+            msg = "{}: {}".format(message_data.formatted_timestamp(), str(message_data).strip())
             with open(self.events_filename, 'a+') as log_file:
-                log_file.write(str(message_data).strip() + '\n')
-            LOGGER.info(message_data)
+                log_file.write(msg + '\n')
+            LOGGER.info(msg)
+
+
+class PrometheusDumper(threading.Thread):
```
Comment (bentsi) [inline sdcm/sct_events.py]: add comment, that it is not used

File: sdcm/sct_events.py [hunk 7]
```diff
@@ -331,30 +388,117 @@ def run(self):
...
+                LOGGER.info(res.text)
+
+    def terminate(self):
+        self.url_set.set()
+        self.stop_event.set()
 
 
 EVENTS_TASK_KILLER = TestKiller()
 EVENTS_FILE_LOOGER = EventsFileLogger()
+EVENTS_PROMETHEUS_DUMPER = PrometheusDumper()
```
Comment (bentsi) [inline sdcm/sct_events.py]: disable

File: sdcm/sct_events.py [hunk 8]
```diff
@@ -331,30 +388,117 @@ def run(self):
...
+                                getattr(message_data, 'type', ''),
+                                message_data.severity,
+                                getattr(message_data, 'node', '')).set(message_data.timestamp)
+
+    def terminate(self):
+        self.stop_event.set()
+
+
+class GrafanaAnnotator(threading.Thread):
+    def __init__(self, specific_dashboard=False):
```
Comment (bentsi) [inline sdcm/sct_events.py]: remove `specific_dashboard`

File: sdcm/sct_events.py [hunk 9]
```diff
@@ -331,30 +388,117 @@ def run(self):
...
+        self.dashboard_id = None
+        self.panel_id = None
+
+        self.auth = ('admin', 'admin')
+        super(GrafanaAnnotator, self).__init__()
+
+    def set_grafana_url(self, grafana_base_url):
+        self.grafana_base_url = grafana_base_url
+        if self.specific_dashboard:
+            res = requests.get(self.grafana_base_url + '/api/search', params={'query': 'nemesis'}, auth=self.auth).json()
```
Comment (bentsi) [inline sdcm/sct_events.py]: separate to functions and remove from the `set_grafana_url` flow

File: sdcm/cluster.py [hunk 10]
```diff
@@ -3096,7 +3099,7 @@ def install_scylla_manager(self, node):
     def set_local_sct_ip(self):
         sct_public_ip = self.params.get('sct_public_ip')
         if sct_public_ip:
-            return sct_public_ip + ':' + self.local_metrics_addr.split(':')[1]
+            return sct_public_ip  # + ':' + self.local_metrics_addr.split(':')[1]
```
[the remark this answers is not in the document]
Comment (fruch) [reply sdcm/cluster.py]: done

File: sdcm/cluster.py (hunk 1 above)
Comment (fruch) [reply sdcm/cluster.py]: done

File: sdcm/cluster.py (hunk 2 above)
Comment (fruch) [reply sdcm/cluster.py]: done

File: sdcm/sct_events.py (hunk 3 above)
Comment (fruch) [reply sdcm/sct_events.py]: done

File: sdcm/sct_events.py (hunk 4 above)
Comment (fruch) [reply sdcm/sct_events.py]: done

File: sdcm/sct_events.py (hunk 6 above)
Comment (fruch) [reply sdcm/sct_events.py]: done

File: sdcm/sct_events.py (hunk 7 above)
Comment (fruch) [reply sdcm/sct_events.py]: done

File: sdcm/sct_events.py (hunk 8 above)
Comment (fruch) [reply sdcm/sct_events.py]: done

File: sdcm/sct_events.py (hunk 9 above)
Comment (fruch) [reply sdcm/sct_events.py]: done

File: sdcm/sct_events.py (hunk 5 above)
Comment (fruch) [reply sdcm/sct_events.py]: done

File: sdcm/cluster.py [hunk 11]
```diff
@@ -3240,14 +3243,27 @@ def configure_alert_manager(self, node):
...
+    @retrying(n=5, sleep_time=10, allowed_exceptions=(Failure, UnexpectedExit),
+              message="Waiting for reconfiguring scylla monitoring")
+    def reconfigure_scylla_monitoring(self):
+        for node in self.nodes:
+            db_targets_list = [n.ip_address for n in self.targets["db_cluster"].nodes]  # node exporter + scylladb
+            self._monitoring_targets = " ".join(db_targets_list)
+            configure_script = dedent("""
+                        cd {0.monitor_install_path}
+                        mkdir -p {0.monitoring_conf_dir}
+                        ./genconfig.py -ns -d {0.monitoring_conf_dir} {0._monitoring_targets}
```
Comment (bentsi) [inline sdcm/cluster.py]: ./genconfig.py -s -n -d

Comment (fruch) [reply sdcm/cluster.py]: done

**Comment by bentsi** [comment]
@fruch please resolve conflicts

**Comment by fruch** [comment]
@bentsi Conflicts are resolved

**Pull request merged**